crates/stages/src/trie/mod.rs

```diff
@@ -0,0 +1,584 @@
+                .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+        }
+
+        let root = H256::from_slice(trie.root().unwrap().as_slice());
+
+        Ok(root)
+    }
+
+    fn calculate_storage_root<DB: Database>(
+        &self,
+        tx: &Transaction<'_, DB>,
+        address: H256,
+    ) -> Result<H256, TrieError> {
+        let db = Arc::new(DupHashDatabase::new(tx, address)?);
+
+        let hasher = Arc::new(HasherKeccak::new());
+
+        let mut trie = PatriciaTrie::new(Arc::clone(&db), Arc::clone(&hasher));
+
+        let mut storage_cursor = tx.cursor_dup_read::<tables::HashedStorage>()?;
+        let mut walker = storage_cursor.walk_dup(address, H256::zero())?;
+
+        while let Some((_, StorageEntry { key: storage_key, value })) = walker.next().transpose()? {
+            let mut out = Vec::new();
+            Encodable::encode(&value, &mut out);
+            trie.insert(storage_key.as_bytes().to_vec(), out)
+                .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+        }
+
+        let root = H256::from_slice(trie.root().unwrap().as_slice());
+
+        Ok(root)
+    }
+
+    /// Calculates the root of the state trie by updating an existing trie.
+    pub(crate) fn update_root<DB: Database>(
+        &self,
+        tx: &Transaction<'_, DB>,
+        root: H256,
+        start_tid: TransitionId,
+        end_tid: TransitionId,
+    ) -> Result<H256, TrieError> {
+        let mut accounts_cursor = tx.cursor_read::<tables::HashedAccount>()?;
+
+        let changed_accounts = self.gather_changes(tx, start_tid, end_tid)?;
+
+        let db = Arc::new(HashDatabase::new_with_root(tx, root)?);
+
+        let hasher = Arc::new(HasherKeccak::new());
+
+        let mut trie = PatriciaTrie::from(Arc::clone(&db), Arc::clone(&hasher), root.as_bytes())
+            .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+
+        for (address, changed_storages) in changed_accounts {
+            if let Some(account) = trie
+                .get(address.as_slice())
+                .map_err(|e| TrieError::InternalError(format!("{e:?}")))?
+            {
+                let storage_root = EthAccount::decode(&mut account.as_slice())?.storage_root;
+                trie.remove(address.as_bytes())
+                    .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+
+                if let Some((_, account)) = accounts_cursor.seek_exact(address)? {
+                    let value = EthAccount::from_with_root(
+                        account,
+                        self.update_storage_root(tx, storage_root, address, changed_storages)?,
+                    );
+
+                    let mut out = Vec::new();
+                    Encodable::encode(&value, &mut out);
+                    trie.insert(address.as_bytes().to_vec(), out)
+                        .map_err(|e| TrieError::InternalError(format!("{e:?}",)))?;
+                }
+            }
```

Comment:
assuming this is expensive

Reply:
Yes, but should be less expensive than rebuilding the full trie when there aren't too many changes, similar to `AccountHashingStage` and `StorageHashingStage`.